src/Event/AccessEventBase.php

```diff
@@ -0,0 +1,98 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\og\Event;
+
+use Drupal\Core\Access\AccessResult;
+use Drupal\Core\Access\AccessResultInterface;
+use Drupal\Core\Cache\RefinableCacheableDependencyInterface;
+use Drupal\Core\Cache\RefinableCacheableDependencyTrait;
+use Drupal\Core\Entity\ContentEntityInterface;
+use Drupal\Core\Session\AccountInterface;
+use Symfony\Component\EventDispatcher\Event;
+
+/**
+ * Base class for OG access events.
+ */
+class AccessEventBase extends Event implements AccessEventInterface {
+
+  use RefinableCacheableDependencyTrait;
+
+  /**
+   * The access result.
+   *
+   * @var \Drupal\Core\Access\AccessResultInterface
+   */
+  protected $access;
+
+  /**
+   * The group that provides the context for the access check.
+   *
+   * @var \Drupal\Core\Entity\ContentEntityInterface
+   */
+  protected $group;
+
+  /**
+   * The user for which to check access.
+   *
+   * @var \Drupal\Core\Session\AccountInterface
+   */
+  protected $user;
+
+  /**
+   * Constructs an AccessEventBase event.
+   *
+   * @param \Drupal\Core\Entity\ContentEntityInterface $group
+   *   The group that provides the context in which to perform the access check.
+   * @param \Drupal\Core\Session\AccountInterface $user
+   *   The user for which to check access.
+   */
+  public function __construct(ContentEntityInterface $group, AccountInterface $user) {
+    $this->group = $group;
+    $this->user = $user;
+    $this->access = AccessResult::neutral();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function grantAccess(): void {
+    $this->access = $this->access->orIf(AccessResult::allowed());
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function denyAccess(): void {
+    $this->access = $this->access->orIf(AccessResult::forbidden());
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getGroup(): ContentEntityInterface {
+    return $this->group;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getUser(): AccountInterface {
+    return $this->user;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getAccessResult(): AccessResultInterface {
+    $access = $this->access;
+
+    if ($access instanceof RefinableCacheableDependencyInterface) {
```

Comment:
Maybe worth adding a comment - I'm not clear on this part?

Reply:
This ensures that our code adheres correctly to our interface and avoids a potential fatal error.

We are returning an `AccessResultInterface` object, and this doesn't offer `::addCacheableDependency()` so we are not supposed to call this method. But in reality all access objects in Drupal core are extending `AccessResult`, which _does_ have this method because it also implements `RefinableCacheableDependencyInterface`. So in typical Drupal code cache metadata is supported and we can call this method. Drupal is full of these kind of discrepancies, and often developers ignore these cases where a method is part of a different interface.

It is not reliable though. It is always possible someone will implement their own `AccessResultInterface` object which does not implement `RefinableCacheableDependencyInterface` and then we get a fatal error. Just putting this simple `if` statement here makes sure we are never going to throw a fatal error if somebody is using a custom implementation.

We in fact are offering extra functionality (cache metadata support) in case the object we are dealing with supports it. This is commonly called "type widening" in PHP. Modern IDE's like PHPStorm have become very good at pointing out these kind of errors.

I will add some documentation here, I will commit it directly in the main branch rather than make a whole PR for it.